Turn context_save/context_restore into standard setjmp/longjmp.

le-jzr opened the pull request:
Since they already _almost_ are identical, this just closes the tiny gap and removes an ugly hack from the standard library in the process.

The assembly changes need to be reviewed by more eyes though.

le-jzr commented on uspace/lib/c/arch/ppc32/src/fibril.S:
```diff
@@ -89,13 +89,13 @@ FUNCTION_BEGIN(context_restore)
-	# context_restore returns 0
-	li r3, 0
+	# __longjmp returns second argument
+	li r3, r4
```
This is probably wrong. `li` stands for load immediate?

jermar replied:
Yeah, very likely. Can't you just mov or move? In the worst case, RTFM :-)

jermar commented on uspace/lib/c/arch/ia64/include/libarch/fibril.h:
```diff
@@ -42,7 +42,7 @@
 
 /*
- * context_save() and context_restore() are both leaf procedures.
+ * setjmp() and __longjmp() are both leaf procedures.
```
The standard says setjmp is a macro.

le-jzr replied:
Ah, indeed, I forgot about that. `#define setjmp(env) setjmp(env)` should do the trick.

jermar commented on uspace/lib/c/arch/amd64/src/fibril.S:
```diff
@@ -53,17 +53,16 @@ FUNCTION_BEGIN(context_save)
 #
 # Restore CPU context from context_t variable
-# pointed by the 1st argument. Returns 0 in EAX.
+# pointed by the 1st argument. Returns RSI in EAX.
```
Maybe rather  "Returns second argument in EAX"

le-jzr commented:
@jermar, I tested it on all emulated platforms except for IA64 (I wasn't able to build ski), which coincidentally is the platform I'm least sure about being correct. :)

jermar commented:
The ia64 changes look good. Ski can often be installed as a package in your favorite Linux distribution.

jxsvoboda commented:
I don't understand why you are renaming it to __setjmp()/__longjmp(). Either they comply with the standard and you would name them setjmp()/longjmp(), or they don't , in which they are not setjmp and longjmp. Why can't they use HelenOS naming conventions then? What's wrong with the names context_save and context_restore then?

le-jzr commented:
That's a good question. Thinking back, there's a history to that. Originally, I was forgetting that `setjmp()` is defined as a macro, so I thought it needs to be a function symbol. That would make it impossible to implement `setjmp()` by wrapping `context_save()` in C, so I renamed `context_save()` to `setjmp()`, so as to avoid more assembly. Normally, I'd change `context_restore()` to `longjmp()`, but that would require an extra conditional branch in assembly for all architectures. Ick. So I made it `__longjmp()`. Then @jermar pointed out that `setjmp()` is only defined as a macro, so I changed the name to `__setjmp()` for consistency. It also expresses that code calling `__setjmp()/__longjmp()` instead of `setjmp()/longjmp()` depends on properties not guaranteed by the standard. But at this point, the motivation for going with the `setjmp`/`longjmp` naming scheme is just cached historical outcome, which I failed to reevaluate.

So that's for the explanation of why. I'll rename them back if you wish.
